File: go/arrays_hashing/0001_two_sum/two_sum.go
package twosum

// TwoSumHashMap finds the indices of two numbers using a HashMap and add up to the target.
// Time Complexity: O(N)
// Space Complexity: O(N)
func TwoSumHashMap(nums []int, target int) []int {
	hash := make(map[int]int)

	for i, num := range nums {
		complement := target - num

		if index, ok := hash[complement]; ok {
			return []int{index, i}
		}

		hash[num] = i
	}

	return nil
}

// TwoSumArray finds the indices of two numbers using a Array Brute Force and add up to the target.
// Time Complexity: O(N^2)
// Space Complexity: O(1)
func TwoSumArray(nums []int, target int) []int {
	for i := 0; i < len(nums); i++ {
		for j := i + 1; j < len(nums); j++ {
			if nums[i]+nums[j] == target {
				return []int{i, j}
			}
		}
	}
	return nil
}

File: python/arrays_hashing/0001_two_sum/two_sum.py
class Solution(object):
    
    def twoSumDictionary(self, nums, target):
        """
        twoSumDictionary finds the indices of two numbers using a Dictionary and add up to the target.
        Time Complexity: O(N)
        Space Complexity: O(N)

        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        numDict = {}

        for i in range(len(nums)):
            complement = target - nums[i]

            if complement in numDict:
                return [numDict[complement], i]
            numDict[nums[i]] = i
        return None
    
    def twoSumArray(self, nums, target):
        """
        twoSumArray finds the indices of two numbers using a Array Brute Force and add up to the target.
        Time Complexity: O(N^2)
        Space Complexity: O(1)

        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i,j]
        return None
